[Merged by Bors] - refactor(geometry/manifold): redefine some instances

urkud opened the pull request:
* Turn `unique_diff_within_at` into a `structure`.
* Generalize `tangent_cone_at`, `unique_diff_within_at`, and `unique_diff_on` to a `semimodule` that is a `topological_space`.
* Redefine `model_with_corners_euclidean_half_space` and `model_with_corners_euclidean_quadrant` to reuse more generic lemmas, including `unique_diff_on.pi` and `unique_diff_on.univ_pi`.
* Make `model_with_corners.unique_diff'` use `target` instead of `range I`; usually it has better defeq.

Co-authored-by: @gebner 

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into real-instance

gebner committed:
perf(analysis/calculus/tangent_cone): make unique_diff_within_at a structure

gebner committed:
tangent_cone: generalize

gebner committed:
Hopefully fix leanchecker issue.

urkud commented on src/geometry/manifold/instances/real.lean:
```diff
@@ -332,33 +264,31 @@ begin
   rcases he with rfl | rfl; rcases he' with rfl | rfl,
-  { -- `e = left chart`, `e' = left chart`
+                 { -- `e = left chart`, `e' = left chart`
     exact (mem_groupoid_of_pregroupoid.mpr (symm_trans_mem_times_cont_diff_groupoid _ _ _)).1 },
```
```suggestion
  { -- `e = left chart`, `e' = left chart`
    exact (mem_groupoid_of_pregroupoid.mpr (symm_trans_mem_times_cont_diff_groupoid _ _ _)).1 },
```

urkud committed:
Update src/geometry/manifold/instances/real.lean

sgouezel commented on src/analysis/calculus/tangent_cone.lean:
```diff
@@ -49,8 +50,9 @@ hence this name. The uniqueness it asserts is proved in `unique_diff_within_at.e
-dense ((submodule.span 𝕜 (tangent_cone_at 𝕜 s x)) : set E) ∧ x ∈ closure s
+@[mk_iff] structure unique_diff_within_at (s : set E) (x : E) : Prop :=
+(dense_tangent_cone : dense ((submodule.span 𝕜 (tangent_cone_at 𝕜 s x)) : set E))
+(mem_closure: x ∈ closure s)
```
```suggestion
(mem_closure : x ∈ closure s)
```

sgouezel commented on src/geometry/manifold/instances/real.lean:
```diff
@@ -82,59 +82,22 @@ a model for manifolds with boundary. In the locale `manifold`, use the shortcut
 def model_with_corners_euclidean_half_space (n : ℕ) [has_zero (fin n)] :
   model_with_corners ℝ (euclidean_space ℝ (fin n)) (euclidean_half_space n) :=
 { to_fun      := λx, x.val,
```
could you also update this one to use the coercion?

urkud replied:
```suggestion
{ to_fun      := coe,
```

sgouezel commented on src/geometry/manifold/instances/real.lean:
```diff
@@ -144,48 +107,17 @@ def model_with_corners_euclidean_quadrant (n : ℕ) :
 { to_fun      := λx, x.val,
```
use the coercion here also?

urkud replied:
```suggestion
{ to_fun      := coe,
```

sgouezel commented on src/geometry/manifold/instances/real.lean:
```diff
@@ -144,48 +107,17 @@ def model_with_corners_euclidean_quadrant (n : ℕ) :
-    ext1 i,
-    simp only [hx i, max_eq_left]
-  end,
+  left_inv'   := λ⟨xval, xprop⟩ hx, by { ext i, simp only [subtype.coe_mk, xprop i, max_eq_left] },
```
```suggestion
  left_inv'   := λ ⟨xval, xprop⟩ hx, by { ext i, simp only [subtype.coe_mk, xprop i, max_eq_left] },
```

sgouezel commented on src/geometry/manifold/instances/real.lean:
```diff
@@ -144,48 +107,17 @@ def model_with_corners_euclidean_quadrant (n : ℕ) :
-    simp only [hx i, max_eq_left]
-  end,
+  left_inv'   := λ⟨xval, xprop⟩ hx, by { ext i, simp only [subtype.coe_mk, xprop i, max_eq_left] },
+  right_inv' := λx hx, by { ext1 i, simp only [hx i, max_eq_left] },
```
```suggestion
  right_inv' := λ x hx, by { ext1 i, simp only [hx i, max_eq_left] },
```

sgouezel reviewed (commented):
LGTM
bors d+
Thanks!

urkud committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

urkud committed:
Merge branch 'real-instance' of git://github.com/leanprover-community/mathlib into real-instance

urkud committed:
Merge branch 'master' into real-instance

urkud committed:
Avoid undefined coercions

There are no coercions from `euclidean_half_space n` or
`euclidean_quadrant n` to `euclidean_space ℝ (fin n)`
